Adding new guides for developing and deploying an ML application in S… (#36700)
ray-project/ray: doc/source/serve/configure-serve-deployment.md

    max_concurrent_queries: 100
    graceful_shutdown_wait_loop_s: 2.0
    graceful_shutdown_timeout_s: 20.0
    health_check_period_s: 10.0
    health_check_timeout_s: 30.0
    ray_actor_options:
      num_cpus: 0.2
      num_gpus: 0.0
```

## Overriding deployment settings

The order of priority is (from highest to lowest):

1. Serve Config file
2. `.options()` call in python code referenced above
3. `@serve.deployment` decorator in python code
4. Serve defaults

For example, if a deployment's `num_replicas` is specified in the config file and their graph code, Serve will use the config file's value. If it's only specified in the code, Serve will use the code value. If the user doesn't specify it anywhere, Serve will use a default (which is `num_replicas=1`).

Keep in mind that this override order is applied separately to each individual parameter.
For example, if a user has a deployment `ExampleDeployment` with the following decorator:

```python
@serve.deployment(
    num_replicas=2,
    max_concurrent_queries=15,
)
class ExampleDeployment:
    ...
```

and the following config file:

```yaml
...

deployments:

    - name: ExampleDeployment
      num_replicas: 5

...
```

> Serve will set `num_replicas=5`, using the config file value, and `max_concurrent_queries=15`, using the code value (since `max_concurrent_queries` wasn't specified in the config file). All other deployment settings use Serve defaults since the user didn't specify them in the code or the config.

Reviewer: ```suggestion
Serve sets `num_replicas=5`, using the config file value, and `max_concurrent_queries=15`, using the code value (because `max_concurrent_queries` wasn't specified in the config file). All other deployment settings use Serve defaults because the user didn't specify them in the code or the config.
```

:::{tip}

> Remember that `ray_actor_options` counts as a single setting. The entire `ray_actor_options` dictionary in the config file overrides the entire `ray_actor_options` dictionary from the graph code. If there are individual options within `ray_actor_options` (e.g. `runtime_env`, `num_gpus`, `memory`) that are set in the code but not in the config, Serve still won't use the code settings if the config has a `ray_actor_options` dictionary. It will treat these missing options as though the user never set them and will use defaults instead. This dictionary overriding behavior also applies to `user_config` and `autoscaling_config`.

Reviewer: ```suggestion
Remember that `ray_actor_options` counts as a single setting. The entire `ray_actor_options` dictionary in the config file overrides the entire `ray_actor_options` dictionary from the graph code. If there are individual options within `ray_actor_options` (e.g. `runtime_env`, `num_gpus`, `memory`) that are set in the code but not in the config, Serve still won't use the code settings if the config has a `ray_actor_options` dictionary. It treats these missing options as though the user never set them and uses defaults instead. This dictionary overriding behavior also applies to `user_config` and `autoscaling_config`.
```

:::

(serve-user-config)=

> ## Dynamically adjusting parameters in deployment (`user_config`)

Reviewer 2: ```suggestion
## Dynamically changing parameters without restarting your replicas (`user_config`)
```

Reviewer 2: long, but more clear

> The `user_config` field can be used to supply structured configuration for your deployment. You can pass arbitrary JSON serializable objects to the YAML configuration. Serve will then apply it to all running and future deployment replicas. The application of user configuration *will not* restart the replica. This means you can use this field to dynamically:

Reviewer: ```suggestion
You can use the `user_config` field to supply structured configuration for your deployment. You can pass arbitrary JSON serializable objects to the YAML configuration. Serve then applies it to all running and future deployment replicas. The application of user configuration *does not* restart the replica. This means you can use this field to dynamically:
```

- adjust model weights and versions without restarting the cluster.
- adjust traffic splitting percentage for your model composition graph.
- configure any feature flag, A/B tests, and hyper-parameters for your deployments.

> To enable the `user_config` feature, you need to implement a `reconfigure` method that takes a dictionary as its only argument:

Reviewer 2: not necessarily a dict, can be any json-serializable object (str or list too)

> To enable the `user_config` feature, you need to implement a `reconfigure` method that takes a json-serializable object (eg. a Dictionary, List or String) as its only argument:

Reviewer: ```suggestion
To enable the `user_config` feature, you need to implement a `reconfigure` method that takes a JSON-serializable object (e.g., a Dictionary, List or String) as its only argument:
```

```python
@serve.deployment
class Model:
    def reconfigure(self, config: Dict[str, Any]):
        self.threshold = config["threshold"]
```

> If the `user_config` is set when the deployment is created (e.g. in the decorator or the Serve config file), this `reconfigure` method is called right after the deployment's `__init__` method, and the `user_config` is passed in as an argument. You can also trigger the `reconfigure` method by updating your Serve config file with a new `user_config` and reapplying it to your Ray cluster. Please see guide for [in-place updates](serve-inplace-updates) for more information.

Reviewer: ```suggestion
If the `user_config` is set when the deployment is created (e.g., in the decorator or the Serve config file), this `reconfigure` method is called right after the deployment's `__init__` method, and the `user_config` is passed in as an argument. You can also trigger the `reconfigure` method by updating your Serve config file with a new `user_config` and reapplying it to your Ray cluster. See [In-place Updates](serve-inplace-updates) for more information.
```

> The corresponding YAML snippet is

Reviewer: ```suggestion
The corresponding YAML snippet is:
```

(serve-configure-deployment)=

# Configure a Ray Serve deployment

These are all the properties available on a Ray Serve deployment. This is also available in the [API reference](../serve/api/doc/ray.serve.deployment_decorator.rst)

> The following can be configured through either the serve config file or on the `@serve.deployment` decorator:

Reviewer 2: A few updates should be made to this content:

- Show a Python code sample and an equivalent config file.
- Show the usage of `.options()`.
- Explain the precedence (config > what's in the code).

I think we should also consolidate the content in the "app builder" page here to make it clear how to compose them. WDYT?

Author: Agreed I'll add this

Author: Added the 3 (actually moved them over from the serve config page because I think they're more relevant here). Also updated the serve config page. For the app builder, do you propose merging app builder content entirely into this page?

- `name` - Name uniquely identifying this deployment within the application. If not provided, the name of the class or function is used.
- `num_replicas` - The number of replicas to run that handle requests to this deployment. Defaults to 1.

> - `route_prefix` - Requests to paths under this HTTP path prefix are routed to this deployment. Defaults to ‘/{name}’. This can only be set for the ingress (top-level) deployment of an application.

Author: I copied this from the application page but is this correct? I think this is supposed to be app_name and not deployment name right? @zcin

> - `route_prefix` - Requests to paths under this HTTP path prefix are routed to this deployment. Defaults to ‘/{name}’. This can only be set for the ingress (top-level) deployment of an application.

Reviewer 3: Actually I think it should default to `/`? Tagging @edoakes because I'm not sure at what point it defaulted to `/{name}`

> - `route_prefix` - Requests to paths under this HTTP path prefix are routed to this deployment. Defaults to ‘/{name}’. This can only be set for the ingress (top-level) deployment of an application.

Reviewer 2: It defaults to `/` now, the `/{name}` is outdated.

> - `route_prefix` - Requests to paths under this HTTP path prefix are routed to this deployment. Defaults to ‘/{name}’. This can only be set for the ingress (top-level) deployment of an application.

Reviewer 2: please fix the docstrings for these too

- `max_concurrent_queries` - The maximum number of queries that are sent to a replica of this deployment without receiving a response. Defaults to 100. This may be an important parameter to check for [performance tuning](serve-perf-tuning).
- `autoscaling_config` - Parameters to configure autoscaling behavior. If this is set, num_replicas cannot be set. For more details on all configurable parameters for autoscaling - [Ray Serve Autoscaling](ray-serve-autoscaling). 
- `health_check_period_s` - How often the health check is called on the replica. Defaults to 10s. The health check is by default a no-op actor call to the replica, but you can define your own as a “check_health” method that raises an exception when unhealthy.
- `health_check_timeout_s` - How long to wait for a health check method to return before considering it failed. Defaults to 30s.
- `graceful_shutdown_wait_loop_s` - Duration that replicas wait until there is no more work to be done before shutting down.

> - `graceful_shutdown_timeout_s` - Duration that a replica can be gracefully shutting down before being forcefully killed.

Author: @edoakes what are the defaults here?

> - `graceful_shutdown_timeout_s` - Duration that a replica can be gracefully shutting down before being forcefully killed.

Reviewer 3: I think it's 2s and 20s: https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/constants.py#L103
